File: tests/test_driver.cpp
#include "mqtt_roof_controller.h"
#include <cassert>
#include <iostream>
#include <memory>
#include <vector>
#include <unordered_map>

// Simple fake client used to verify that commands are published to the
// expected topics.
class FakeMqttClient : public IMqttClient {
public:
    void connect() override {}
    void disconnect() override {}
    void publish(const std::string& topic, const std::string& payload) override {
        published.emplace_back(topic, payload);
    }
    void subscribe(const std::string& topic,
                   std::function<void(const std::string&)> cb) override {
        subs[topic] = std::move(cb);
    }

    void emit(const std::string& topic, const std::string& payload) {
        if (subs.count(topic)) subs[topic](payload);
    }

    std::vector<std::pair<std::string, std::string>> published;
    std::unordered_map<std::string, std::function<void(const std::string&)>> subs;
};

int main() {
    auto client = std::make_shared<FakeMqttClient>();
    MQTTRoofController controller(client,
                                 "roof/open",
                                 "roof/close",
                                 "roof/open_limit",
                                 "roof/close_limit",
                                 "roof/percent",
                                 "roof/power");
    controller.connect();

    controller.set_power(true);
    controller.open_roof();
    controller.close_roof();
    controller.set_open_percentage(42);

    assert(client->published.size() == 4);
    assert(client->published[0] == std::make_pair(std::string("roof/power"), std::string("1")));
    assert(client->published[1] == std::make_pair(std::string("roof/open"), std::string("1")));
    assert(client->published[2] == std::make_pair(std::string("roof/close"), std::string("1")));
    assert(client->published[3] == std::make_pair(std::string("roof/percent"), std::string("42")));

    // Simulate limit switch callbacks
    client->emit("roof/open_limit", "1");
    client->emit("roof/close_limit", "1");
    assert(controller.is_open());
    assert(controller.is_closed());

    std::cout << "All tests passed\n";
    return 0;
}

File: tests/test_driver.py
import os
import sys
import types
from unittest.mock import MagicMock

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))

from roro_mqtt.driver import MQTTRoofController, configure_topics


def make_msg(topic, payload):
    return types.SimpleNamespace(topic=topic, payload=payload)


def test_publish_commands():
    configure_topics()  # reset defaults
    client = MagicMock()
    driver = MQTTRoofController(
        host="localhost",
        client=client,
        topic_open="open",
        topic_close="close",
        topic_power="power",
    )

    driver.open_roof()
    client.publish.assert_any_call("open", payload="1")

    driver.close_roof()
    client.publish.assert_any_call("close", payload="1")

    driver.set_power(True)
    client.publish.assert_any_call("power", payload="ON")


def test_state_updates():
    configure_topics()  # reset defaults
    client = MagicMock()
    driver = MQTTRoofController(
        host="localhost",
        client=client,
        topic_open_limit="ol",
        topic_close_limit="cl",
        topic_percent="percent",
    )

    driver._on_message(client, None, make_msg("ol", b"1"))
    assert driver.open_limit is True

    driver._on_message(client, None, make_msg("cl", b"1"))
    assert driver.close_limit is True

    driver._on_message(client, None, make_msg("percent", b"45"))
    assert driver.percent_open == 45.0


def test_configure_topics():
    client = MagicMock()
    configure_topics(open="cfg/open", close="cfg/close", power="cfg/power")
    driver = MQTTRoofController(host="localhost", client=client)

    driver.open_roof()
    client.publish.assert_any_call("cfg/open", payload="1")

    driver.close_roof()
    client.publish.assert_any_call("cfg/close", payload="1")

    driver.set_power(True)
    client.publish.assert_any_call("cfg/power", payload="ON")

    # Reset for other tests
    configure_topics()
